File: R/day2.R
library(tidyverse)
library(widyr)

day2 <- read_table("data/day2.txt", col_names = F) %>%
  select(pattern = X1)

checksum <- function(x, n) {
  y <- x %>%
    str_split("", simplify = T) %>%
    table() %>%
    as.vector()
  return(sum(any(y == n)))
}

day2 %>%
  mutate(
    two = map_lgl(pattern, checksum, n = 2),
    three = map_lgl(pattern, checksum, n = 3)
  ) %>%
  summarize_at(vars(-starts_with("pattern")), sum) %>%
  mutate(
    product = two * three
  )

# Part 2

part2 <- day2 %>% pull(pattern)

common_chars <- function(x) {
  hamming <- stringdist::stringdistmatrix(x, x, "hamming")
  correct_idx <- which(hamming == 1) %% length(x) 
  # return(correct_idx)
  correct <- x[correct_idx] %>% str_split("")
  return(glue::glue_collapse(correct[[1]][-which(correct[[1]] != correct[[2]])]) %>% as.character())
}

# sometimes tidy methods dont work
# crossing(item1 = test, item2 = test) %>%
#   mutate(
#     hamming = map2_dbl(item1, item2, stringdist::stringdist, method = "hamming")
#   ) %>%
#   filter(hamming == 1) %>%
#   slice(1) %>%
#   mutate(
#     common = map2(item1, item2, function(x, y) {
#       c(x, y) %>% 
#         str_split("") 
#     })
#   )

File: python/day2.py
c2 = 0
c3 = 0
strings = []
for line in open("data/day2.txt"):
  s = line.strip()
  print(s)
  strings.append(s)
  count = {}
  for c in s:
    if c not in count:
      count[c] = 0
    count[c] += 1
  two = False
  three = False
  for k, v in count.items():
    if v == 2:
      two = True
    if v == 3:
      three = True

  if two:
    c2 += 1
  if three:
    c3 += 1

print(c2 * c3)

for s1 in strings:
  for s2 in strings:
    diff = 0
    for i in range(len(s1)):
      if s1[i] != s2[i]:
        diff += 1
    if diff == 1:
      final = []
      for i in range(len(s1)):
        if s1[i] == s2[i]:
          final.append(s1[i])
      print(''.join(final)) 
      print(s1, s2)